### PR TITLE
Create Detector Workflow

### PR DESCRIPTION
**Pull request by owaiskazi19**
### Description
<s>This is an initial draft PR for create detector workflow. 
1. When creating an [index](https://github.com/opensearch-project/anomaly-detection/blob/main/src/main/java/org/opensearch/ad/indices/AnomalyDetectionIndices.java#L538) and having the method markMappingUpToDate which requires an actionListener. I have passed it as null for now but we need to figure out a way for this mapping. Or do we even need a listener because we can get the Response directly while creating the AD_Index from SDKClient.
2. Need to wait for createComponent [integration](https://github.com/opensearch-project/anomaly-detection/pull/684) to finish so that we will get an object for AnomalyDetectionIndices to call markMappingUpToDate
3. Not able to start AD Extension because while creating the handler I'm passing null for settings and clusterService
 `handler = List.of(new RestCreateDetectorAction(null, null)); ` as these objects will be available through createComponent </s>

This PR adds:
1. Create Detector workflow for AD Extension
2. Migrate client transports to Apache HttpClient / Core 5.x

### Issues Resolved
https://github.com/opensearch-project/opensearch-sdk-java/issues/58

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java [hunk 1]
```diff
@@ -39,7 +44,14 @@ public ExtensionSettings getExtensionSettings() {
 
     @Override
     public List<ExtensionRestHandler> getExtensionRestHandlers() {
-        return List.of(new RestCreateDetectorAction());
+        List<ExtensionRestHandler> handler = null;
+        // Get the settings anc clusterService from createComponent
+        try {
+            handler = List.of(new RestCreateDetectorAction(null, null));
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java:50]: First null is `Settings`, second is cluster service.  
Settings should be available via `ExtensionsRunner.sendEnvironmentSettingsRequest()`.  

BaseExtension createComponents already stores cluster state, just need to actually call that from the ExtensionsRunner at some point, and also add settings to that.

File: src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java [hunk 2]
```diff
@@ -39,7 +44,14 @@ public ExtensionSettings getExtensionSettings() {
...
+        } catch (IOException e) {
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java:51]: does instantiating this object throw an exception?

File: src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java [hunk 3]
```diff
@@ -83,6 +95,11 @@ public List<Setting<?>> getSettings() {
         );
     }
 
+    @Override
+    public Collection<Object> createComponents(SDKClient sdkClient, ClusterService clusterService, ThreadPool threadPool) {
+        return null;
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java:100]: Not sure this is ever read anywhere, but isn't returning an empty collection better than null?

We don't need to override this if we implement in BaseExtension.

File: src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java [hunk 4]
```diff
@@ -3,18 +3,23 @@
...
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import org.opensearch.ad.rest.RestCreateDetectorAction;
 import org.opensearch.ad.settings.AnomalyDetectorSettings;
 import org.opensearch.ad.settings.EnabledSetting;
+import org.opensearch.client.opensearch.OpenSearchClient;
+import org.opensearch.cluster.service.ClusterService;
 import org.opensearch.common.settings.Setting;
+import org.opensearch.sdk.*;
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java:17]: Given how much confusion I know was generated in creating this PR with similarly-named classes, this star import should be a big red flag. :-)

File: src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java [hunk 5]
```diff
@@ -39,7 +44,13 @@ public ExtensionSettings getExtensionSettings() {
 
     @Override
     public List<ExtensionRestHandler> getExtensionRestHandlers() {
-        return List.of(new RestCreateDetectorAction());
+        List<ExtensionRestHandler> handler = null;
+        try {
+            handler = List.of(new RestCreateDetectorAction());
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java:49]: We don't need a catch block for this object instantiation.  The constructor is a no-op.

File: src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java [hunk 6]
```diff
@@ -91,6 +107,12 @@ private static ExtensionSettings initializeSettings() throws IOException {
         return settings;
     }
 
+    public OpenSearchClient getClient() throws IOException {
+        SDKClient sdkClient = new SDKClient();
+        OpenSearchClient client = sdkClient.initializeClient("localhost", 9200);
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java:112]: Let's see if we can fetch these parameters from somewhere else.

File: src/main/java/org/opensearch/ad/rest/RestCreateDetectorAction.java [hunk 7]
```diff
@@ -1,134 +1,241 @@
...
-import static org.opensearch.rest.RestRequest.Method.GET;
-import static org.opensearch.rest.RestRequest.Method.PUT;
-import static org.opensearch.rest.RestStatus.BAD_REQUEST;
-import static org.opensearch.rest.RestStatus.NOT_ACCEPTABLE;
-import static org.opensearch.rest.RestStatus.NOT_FOUND;
-import static org.opensearch.rest.RestStatus.NOT_MODIFIED;
+import static org.opensearch.ad.model.AnomalyDetector.ANOMALY_DETECTORS_INDEX;
+import static org.opensearch.ad.settings.AnomalyDetectorSettings.ANOMALY_DETECTORS_INDEX_MAPPING_FILE;
+import static org.opensearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+import static org.opensearch.rest.RestRequest.Method.*;
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/ad/rest/RestCreateDetectorAction.java:6]: I love static imports but not ones with *.

File: src/main/java/org/opensearch/ad/rest/RestCreateDetectorAction.java [hunk 8]
```diff
@@ -1,134 +1,241 @@
...
-    @Override
-    public ExtensionRestResponse handleRequest(ExtensionRestRequest request) {
-        Method method = request.method();
+    private String getAnomalyDetectorMappings() throws IOException {
+        URL url = AnomalyDetectionIndices.class.getClassLoader().getResource(ANOMALY_DETECTORS_INDEX_MAPPING_FILE);
+        return Resources.toString(url, Charsets.UTF_8);
+    }
+
+    private IndexResponse indexAnomalyDetector(AnomalyDetector anomalyDetector) throws IOException {
+        AnomalyDetector detector = new AnomalyDetector(
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/ad/rest/RestCreateDetectorAction.java:77]: It looks here like we're pretty much cloning the input parameter.  Why is this needed?

File: src/main/java/org/opensearch/ad/rest/RestCreateDetectorAction.java [hunk 9]
```diff
@@ -1,134 +1,241 @@
...
+    @Override
+    public ExtensionRestResponse handleRequest(ExtensionRestRequest request) {
+        if (!EnabledSetting.isADPluginEnabled()) {
+            throw new IllegalStateException(CommonErrorMessages.DISABLED_ERR_MSG);
+        }
+        Method method = request.method();
+
+        NamedXContentRegistry xContentRegistry = new NamedXContentRegistry(getNamedXWriteables());
+        XContentParser parser = null;
+        try {
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/ad/rest/RestCreateDetectorAction.java:191]: I think we can try to wrap most of this handler method in a try/catch, where the catch returns an ExtensionRestResponse failure with the exception details.

File: src/main/java/org/opensearch/ad/rest/RestCreateDetectorAction.java (hunk 8 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/ad/rest/RestCreateDetectorAction.java:75]: IndexRequest requires a model class object to document it with the index.
```
IndexRequest<AnomalyDetector> indexRequest = new IndexRequest.Builder<AnomalyDetector>()
            .index(ANOMALY_DETECTORS_INDEX)
            .id("1")
            .document(detector)
            .build();

File: src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java (hunk 6 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java:106]: Raised https://github.com/opensearch-project/opensearch-sdk-java/pull/207 for the same

File: src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java (hunk 1 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java]: No need of settings and clusterService anymore. Removed it :)

File: src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java (hunk 2 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java]: Removed the exception from SDK: https://github.com/opensearch-project/opensearch-sdk-java/pull/234

File: src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java (hunk 3 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java:93]: Not extending from `AbstractAnomalyDetectorAction` anymore. So not needed, removed!

File: src/main/java/org/opensearch/ad/model/AnomalyDetector.java [hunk 10]
```diff
@@ -65,15 +65,14 @@
...
-
     public static final String PARSE_FIELD_NAME = "AnomalyDetector";
     public static final NamedXContentRegistry.Entry XCONTENT_REGISTRY = new NamedXContentRegistry.Entry(
         AnomalyDetector.class,
         new ParseField(PARSE_FIELD_NAME),
         it -> parse(it)
     );
     public static final String NO_ID = "";
-    public static final String ANOMALY_DETECTORS_INDEX = ".opendistro-anomaly-detectors";
+    public static final String ANOMALY_DETECTORS_INDEX = "opendistro-anomaly-detectors";
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/ad/model/AnomalyDetector.java:75]: (Random thought, don't hold up PR approval for this)

Do we want to keep the opendistro name for backwards compatibility?  Although removing the dot (per 3.x) breaks that anyway, so we might want a new name.

File: src/main/java/org/opensearch/ad/rest/RestCreateDetectorAction.java [hunk 11]
```diff
@@ -1,134 +1,227 @@
...
-        String worldNameWithRandomAdjective = worldAdjectives.isEmpty()
-            ? worldName
-            : String.join(" ", worldAdjectives.get(rand.nextInt(worldAdjectives.size())), worldName);
-        return new ExtensionRestResponse(request, OK, String.format(GREETING, worldNameWithRandomAdjective));
+    public List<NamedXContentRegistry.Entry> getNamedXWriteables() {
+        List<NamedXContentRegistry.Entry> entries = new ArrayList<>();
+        entries.add(AnomalyDetector.XCONTENT_REGISTRY);
+        entries.add(AnomalyResult.XCONTENT_REGISTRY);
+        entries.add(DetectorInternalState.XCONTENT_REGISTRY);
+        entries
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/ad/rest/RestCreateDetectorAction.java:110]: Note to other reviewers: these four `entries.add()` here and below are placeholder code and will be replaced by getNamedXContent work in progress.

This comment may be resolved when ready to merge.

**Review by dbwiddis — APPROVED** [review]
LGTM! Great job! A few comments (can be resolved to merge).

File: src/main/java/org/opensearch/ad/model/AnomalyDetector.java (hunk 10 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/ad/model/AnomalyDetector.java:75]: This is really good point. Do you think we should discuss this on an issue on SDK repo or if you have a better name I can replace it in this PR itself?
I can think of `opensearch-anomaly-detector`

Comment (owaiskazi19) [reply src/main/java/org/opensearch/ad/model/AnomalyDetector.java:75]: Actually, let me open an issue for this because I think it will require some discussion

Comment (owaiskazi19) [reply src/main/java/org/opensearch/ad/model/AnomalyDetector.java:75]: https://github.com/opensearch-project/opensearch-sdk-java/issues/241

**Review by ryanbogan — APPROVED** [review]
